src/solver/impls/snes/snes.cxx

```diff
@@ -21,153 +21,262 @@
+    BoutReal target = simtime + out_timestep;
+
+    bool looping = true;
+    do {
+      // Copy the state (snes_x) into initial values (x0)
+      VecCopy(snes_x, x0);
+
+      // Set the timestep
+      dt = timestep;
+      looping = true;
+      if (simtime + dt >= target) {
+        // Ensure that the timestep goes to the next output time and then stops
+        looping = false;
+        dt = target - simtime;
+      }
+
+      if (predictor and (time1 > 0.0)) {
+        // Use (time1, x1) and (simtime, x0) to make prediction
+        // snes_x <- x0 + (dt / (simtime - time1)) * (x0 - x1)
+        // snes_x <- -β * x1 + (1 + β) * snes_x
+        BoutReal beta = dt / (simtime - time1);
+        VecAXPBY(snes_x, -beta, (1. + beta), x1);
+      }
+
+      // Run the solver
+      SNESSolve(snes, nullptr, snes_x);
+
+      // Find out if converged
+      SNESConvergedReason reason;
+      SNESGetConvergedReason(snes, &reason);
+      if (reason < 0) {
+        // Diverged
+
+        // Try a smaller timestep
+        timestep /= 2.0;
+        // Restore state
+        VecCopy(x0, snes_x);
+
+        // Check lock state
+        PetscInt lock_state;
+        VecLockGet(snes_x, &lock_state);
+        if (lock_state > 0) {
+          // Locked for read
+          output.write("WARNING: snes_x locked for reading\n");
+        } else if (lock_state < 0) {
+          // Locked for write
+          output.write("WARNING: snes_x locked for writing\n");
+        }
+        continue; // Try again
+      }
+
+      if (predictor) {
+        // Save previous values: x1 <- x0
+        VecCopy(x0, x1);
+        time1 = simtime;
+      }
+
+      simtime += dt;
+      int its;
+      SNESGetIterationNumber(snes, &its);
+
+      if (diagnose) {
+        output.write("SNES time: {}, timestep: {}, iterations: {}\n", simtime, timestep,
+                     its);
+      }
+
+      if (looping) {
+        if (its <= lower_its) {
+          // Increase timestep slightly
+          timestep *= 1.1;
+        } else if (its >= upper_its) {
+          // Reduce timestep slightly
+          timestep *= 0.9;
+        }
+      }
+    } while (looping);
+
+    // Put the result into variables
+    {
+      const BoutReal* xdata = nullptr;
+      int ierr = VecGetArrayRead(snes_x, &xdata);
+      CHKERRQ(ierr); // NOLINT
+      load_vars(const_cast<BoutReal*>(xdata));
+      ierr = VecRestoreArrayRead(snes_x, &xdata);
+      CHKERRQ(ierr); // NOLINT
+    }
+    run_rhs(simtime); // Run RHS to calculate auxilliary variables
+
+    /// Call the monitor function
+
+    if (call_monitors(simtime, s, nsteps) != 0) {
+      break; // User signalled to quit
+    }
   }
-    
+
   return 0;
 }
 
 // f = rhs
 PetscErrorCode SNESSolver::snes_function(Vec x, Vec f) {
-  const BoutReal *xdata;
-  BoutReal *fdata;
-  int ierr;
-
   // Get data from PETSc into BOUT++ fields
-  ierr = VecGetArrayRead(x,&xdata);CHKERRQ(ierr);
+  const BoutReal* xdata = nullptr;
```

Comment:
again `const`?

Reply:
just for `VecGetArrayRead`. Since we're reading not writing I think it's best to make it const, though `load_vars` needs the const casting away